README.md
# OSTicket

Wrapper for OSTicket. At the moment you can only create tickets.

***ATTENTION/BREAKING CHANGES in version 2***
+ no supports for callbacks. Please use async/await
+ response is an object (for more flexibility) with property ticketId instead of a plain string

# Installation
```
  yarn add ac-osticket
  npm install ac-osticket --save
```

# Usage
Create API Key in OSTicket and instanciate OSTicket with your OSTicket URL and those API keys.

## Init
```
const ost = require('osticket')
ost.init({
  baseUrl: 'https://myOSTicketURL',
  apiKey: 'abc-123', 
  apiSecret: 'abc-secret' // optional if you have tweaked OSTicket
})

// see below for payload
let response = await ost.createTicket(ticket)
{ ticketId: 123456 }
```

### Init params
| Property | Type | Notes |
| --- | --- | --- |
| baseUrl | string | required base url of your OS ticket instance |
| apiKey | string | required API key |
| apiSecret | string | optional, see Tweak section below
| debugMode | boolean | If true, no real tickets are created
| keylock | object | Optional parameters for keylock (e.g. redis to use Redis instead of local memory) 


## Create a New Ticket
```
const ticket = {
    "name": "Jane Doe",
    "email": "jane.doe@admiralcloud.com",
    "subject": "I need help",
    "message": "This is my bug report"
}

let response = await ost.createTicket(ticket)
```

Optionally you can use our locking mechanism to block creating too many tickets.

```
let options = {
    key: 'ticketLock', // name for the key to use as lock key
    expires: 5 // time to block in seconds
}
let response = await ost.createTicket(ticket, options)
```

## Custom fields
You can also send custom fields to the OSTicket API. Make sure to add definitions (based on ac-sanitizer).

```
const ticket = {
    name: 'Jane Doe',
    email: 'jane.doe@admiralcloud.com',
    subject: 'I need help',
    message: 'This is my bug report',
    jobNumber: 'ABC-JobId-123'
}

const options = {
    fieldsToCheck: [
        { field: 'jobNumber', type: 'string', required: true },
    ]
}
let response = await ost.createTicket(ticket, options)

```


## Test mode
You can enable test mode, which will not create real tickets but instead return a random ID and debugMode true in the response

```
{
    ticketId: 123456,
    debugMode: true
}
```

Enable testmode globally by using init parameter "debugmode" to true or use it on per-function basis and create a ticket with option debug = true.

```
ost.init({
  baseUrl: 'https://myOSTicketURL',
  apiKey: 'abc-123', 
  apiSecret: 'abc-secret' // optional if you have tweaked OSTicket,
  debugMode: true
})

// OR
const ticket = {
    "name": "Jane Doe",
    "email": "jane.doe@admiralcloud.com",
    "subject": "I need help",
    "message": "This is my bug report"
}

let response = await ost.createTicket(ticket, { debug: true })
```

# OSTicket Tweaks
API Keys are currently bound to IP addresses. This might be a problem if you have more than one IP and - due to scaling - might always have several and different IPs. Therefore the OSTicket is slightly modified to handle this.

We are using a "secret" header to authenticate the API, additionally to the access key. 

Please go to  include/class.api.php in your OSTicket installation and scroll to the function requireApiKey(). Update this function and the function getApiKey() according to the code below. This is not a kryptographically safe approach, as the header is visible for everyone. But currently there is no "safe" approach. 

```
// Line 54 in include/class.api.php
function getSecret() {
    return $this->ht['apisecret'];
}
 

// Line 106 in include/class.api.php
function getIdByKey($key, $ip='', $secret='') {
    $sql='SELECT id FROM '.API_KEY_TABLE.' WHERE apikey='.db_input($key);
    if($ip)
        $sql.=' AND ipaddr='.db_input($ip);
    if($secret)
        $sql.=' AND apisecret='.db_input($secret);
    if(($res=db_query($sql)) && db_num_rows($res))
        list($id) = db_fetch_row($res);

    return $id;
}

// Line 178 in include/class.api.php
function requireApiKey() {
    # Validate the API key -- required to be sent via the X-API-Key
    # header
    # CHANGED: Mark Poepping/AdmiralCloud AG 2017-06-09
    # If APIsecret is sent, don't check requirement for IP

    if(!($key=$this->getApiKey()))
        return $this->exerr(401, __('Valid API key required'));
    else if (!$key->isActive())
        return $this->exerr(401, __('API key not active'));         
    else if (!$key->getSecret() && $key->getIPAddr()!=$_SERVER['REMOTE_ADDR'])
        return $this->exerr(401, __('API key - IP Restriction or API secret must be set'));

    return $key;
}
 
 
// Line 193 in include/class.api.php
function getApiKey() {
    if (!$this->apikey && isset($_SERVER['HTTP_X_API_KEY']) && isset($_SERVER['HTTP_X_API_AUTH']))
        $this->apikey = API::lookupByKey($_SERVER['HTTP_X_API_KEY'], '', $_SERVER['HTTP_X_API_AUTH']);
    else if (!$this->apikey && isset($_SERVER['HTTP_X_API_KEY']) && isset($_SERVER['REMOTE_ADDR']))
        $this->apikey = API::lookupByKey($_SERVER['HTTP_X_API_KEY'], $_SERVER['REMOTE_ADDR']);
         
    return $this->apikey;
}
 
 
// This pull request
// https://github.com/osTicket/osTicket/pull/4281/commits/a23e092941ab2c27bf0de4a1cfbbcd6976a5d423
```


# License & Contributing

## Links
- [Website](https://www.admiralcloud.com/)
- [Twitter (@admiralcloud)](https://twitter.com/admiralcloud)
- [Facebook](https://www.facebook.com/MediaAssetManagement/)

## License
[MIT License](https://opensource.org/licenses/MIT) Copyright © 2009-present, AdmiralCloud AG, Mark Poepping

## Contributing
If you want to contribute, please make sure to add tests and have code coverage of 100% when running *yarn run test*.

## Thanks
This module is inspired by https://github.com/hongkongkiwi/osticket-node and https://github.com/kumarharsh/node-freshdesk



CHANGELOG.md
<a name="2.0.4"></a>

## [2.0.4](https://github.com/mmpro/ac-osticket/compare/v2.0.3..v2.0.4) (2021-05-01 13:39:21)


### Bug Fix

* **App:** Allow parameter keylock for init | MP | [5a9817ad856b3c3f48a2e31aa3fdbff45692f0c6](https://github.com/mmpro/ac-osticket/commit/5a9817ad856b3c3f48a2e31aa3fdbff45692f0c6)    
Allow parameter keylock for init
<a name="2.0.3"></a>

## [2.0.3](https://github.com/mmpro/ac-osticket/compare/v2.0.2..v2.0.3) (2021-05-01 13:14:58)


### Bug Fix

* **App:** Response is always an object | MP | [04fdb81f9436b4871b171e0d606fcd4f35e880d8](https://github.com/mmpro/ac-osticket/commit/04fdb81f9436b4871b171e0d606fcd4f35e880d8)    
Response is always an object
### Tests

* **App:** Added tests with custom fields | MP | [f49a52e6162a8b21d66b85557dce7162be2179ce](https://github.com/mmpro/ac-osticket/commit/f49a52e6162a8b21d66b85557dce7162be2179ce)    
Added tests with custom fields
### Documentation

* **App:** Updated README | MP | [7803066bfe2ca3b2c5910e2072fe97f7cdcb8029](https://github.com/mmpro/ac-osticket/commit/7803066bfe2ca3b2c5910e2072fe97f7cdcb8029)    
Updated README
<a name="2.0.2"></a>

## [2.0.2](https://github.com/mmpro/ac-osticket/compare/v2.0.1..v2.0.2) (2021-05-01 11:13:49)


### Bug Fix

* **App:** Allow debug mode | MP | [4026d40973895893cd64f5ab7a67260336c2405b](https://github.com/mmpro/ac-osticket/commit/4026d40973895893cd64f5ab7a67260336c2405b)    
Allow debug mode
<a name="2.0.1"></a>

## [2.0.1](https://github.com/mmpro/ac-osticket/compare/v2.0.0..v2.0.1) (2021-05-01 07:30:53)


### Bug Fix

* **App:** Use apiKey (camel-cased) for consistency | MP | [68cf23cdf9872deacd17b23d240c233c5e2f7b73](https://github.com/mmpro/ac-osticket/commit/68cf23cdf9872deacd17b23d240c233c5e2f7b73)    
As all parameters are camel-cased apiKey is now also camel-cased (apiKey instead of apikey)
### Documentation

* **App:** Added OSTicket tweak for usage with secrets | MP | [d90ea3bc8946b5018c63aa543ba5dde4aa34a632](https://github.com/mmpro/ac-osticket/commit/d90ea3bc8946b5018c63aa543ba5dde4aa34a632)    
Added OSTicket tweak for usage with secrets
### Chores

* **App:** Add nyc for code coverage | MP | [cede2c785be3b0bbd77c7e78e4226198edea4f41](https://github.com/mmpro/ac-osticket/commit/cede2c785be3b0bbd77c7e78e4226198edea4f41)    
Add nyc for code coverage
<a name="2.0.0"></a>
 
# [2.0.0](https://github.com/mmpro/ac-osticket/compare/v1.0.5..v2.0.0) (2021-05-01 06:52:24)


### Feature

* **App:** New version with async/await | MP | [1fbd824f4ea966e0c2b7d5d97660b17f496a5998](https://github.com/mmpro/ac-osticket/commit/1fbd824f4ea966e0c2b7d5d97660b17f496a5998)    
Re-factored module which now supports async/await but not longer uses/supports callbacks.
## BREAKING CHANGES
* **App:** New version with async/await
<a name="1.0.5"></a>

## [1.0.5](https://github.com/mmpro/ac-osticket/compare/v1.0.4..v1.0.5) (2021-04-12 17:02:02)


### Bug Fix

* **App:** Minor fix | MP | [7defe194d9f7a7664bf105464f93d38efc260fe9](https://github.com/mmpro/ac-osticket/commit/7defe194d9f7a7664bf105464f93d38efc260fe9)    
Minor fix
### Chores

* **App:** Updated packages | MP | [c1be041336ae0ce19aca200ed1b96b02999e216f](https://github.com/mmpro/ac-osticket/commit/c1be041336ae0ce19aca200ed1b96b02999e216f)    
Updated packages
<a name="1.0.4"></a>

## [1.0.4](https://github.com/mmpro/ac-osticket/compare/v1.0.3..v1.0.4) (2020-03-29 12:52:39)


### Bug Fix

* **App:** Maintenance release | MP | [f187339fe62212cff78aeaf8f016f2cc70dec29b](https://github.com/mmpro/ac-osticket/commit/f187339fe62212cff78aeaf8f016f2cc70dec29b)    
Maintenance release
* **App:** Maintenance release | MP | [b61d8610c31b9861860c985282d5e4696b0e43f1](https://github.com/mmpro/ac-osticket/commit/b61d8610c31b9861860c985282d5e4696b0e43f1)    
Maintenance release to make sure installation from NPM is working (last version there is 1.0.2)
### Chores

* **App:** Updated packages | MP | [6bde0783b4c4eb2325af50c2a5ae5beb68b43ed4](https://github.com/mmpro/ac-osticket/commit/6bde0783b4c4eb2325af50c2a5ae5beb68b43ed4)    
Updated packages
* **release:** v0.0.3 [ci skip] | [e1daf485d0cde9e2969255e1954e2fdea9551189](https://github.com/mmpro/ac-osticket/commit/e1daf485d0cde9e2969255e1954e2fdea9551189)    

* **release:** v0.0.2 [ci skip] | [c770e369f706b13c214f888632474581a8c34605](https://github.com/mmpro/ac-osticket/commit/c770e369f706b13c214f888632474581a8c34605)    

<a name="0.0.3"></a>

## [0.0.3](https://github.com/mmpro/ac-osticket/compare/v0.0.2..v0.0.3) (2020-03-29 12:46:58)


### Bug Fix

* **App:** Maintenance release | MP | [f187339fe62212cff78aeaf8f016f2cc70dec29b](https://github.com/mmpro/ac-osticket/commit/f187339fe62212cff78aeaf8f016f2cc70dec29b)    
Maintenance release
<a name="0.0.2"></a>

## [0.0.2](https://github.com/mmpro/ac-osticket/compare/v0.0.1..v0.0.2) (2020-03-29 12:46:03)


### Bug Fix

* **App:** Maintenance release | MP | [b61d8610c31b9861860c985282d5e4696b0e43f1](https://github.com/mmpro/ac-osticket/commit/b61d8610c31b9861860c985282d5e4696b0e43f1)    
Maintenance release to make sure installation from NPM is working (last version there is 1.0.2)
<a name="0.0.1"></a>

## [0.0.1](https://github.com/mmpro/ac-osticket/compare/..v0.0.1) (2020-03-29 12:42:09)


### Bug Fix

* **App:** Prepare repository for AC semantic release | MP | [c425b7eaf4edd4788fbd8957e28f43bda18af9d4](https://github.com/mmpro/ac-osticket/commit/c425b7eaf4edd4788fbd8957e28f43bda18af9d4)    
Cleaned up repository and use ac-semantic-release
### Documentation

* **App:** Minor changes in README | MP | [c34610eeae967606b36912a244a3a53c7a14443c](https://github.com/mmpro/ac-osticket/commit/c34610eeae967606b36912a244a3a53c7a14443c)    
Minor changes in README
### Style

* **App:** Lint fix | MP | [3fd0e1a0e4ead56055ad75a10aa48801701a3c33](https://github.com/mmpro/ac-osticket/commit/3fd0e1a0e4ead56055ad75a10aa48801701a3c33)    
Lint fix
### Chores


test.js
/**
 * PREPARE HTTP MOCK SERVER AS OSTICKET ENDPOINT
 */
const _ = require('lodash')
const http = require('http')

/**
 * 
Status: 201 Created
123456
 */
const requestListener = (req, res) => {
  res.statusCode = 201
  let data = '';
  req.on('data', chunk => {
    data += chunk;
  })
  req.on('end', () => {
    data = JSON.parse(data.toString())
    const ticketId = Math.floor(Math.random() * 100000)
    if (_.get(data, 'payloadCheck')) {
      const payload = Buffer.from(JSON.stringify(data)).toString('base64')
      res.setHeader('x-ac-payloadhash', payload)  
    }
    res.end(ticketId.toString())
  })
}
const server = http.createServer(requestListener);
server.listen(8070)


/**
 * TEST SECTION
 */

const expect = require('chai').expect
const ost = require('./index')
ost.init({
  baseUrl: 'http://localhost:8070/',
  apiKey: 'abc-123', 
  apiSecret: 'abc-secret' 
})


describe('Run test', function() {
  this.timeout(15000)

  it('Should create a ticket', async() => {
    const ticket = {
      name: 'Jane Doe',
      email: 'jane.doe@admiralcloud.com',
      subject: 'I need help',
      message: 'This is my bug report'
    }
    let response = await ost.createTicket(ticket)
    expect(response.ticketId).to.be.a('number')
  })

  it('Try to create a ticket with missing email - should fail', async() => {
    const ticket = {
      name: 'Jane Doe',
      subject: 'I need help',
      message: 'This is my bug report'
    }
    let response = await ost.createTicket(ticket)
    expect(response).to.have.property('message', 'field_email_required')
  })

  it('Should create a ticket and set lock for 5 seconds', async() => {
    const ticket = {
      name: 'Jane Doe',
      email: 'jane.doe@admiralcloud.com',
      subject: 'I need help',
      message: 'This is my bug report'
    }
    const options = {
      key: 'ticketLock',
      expires: 5
    }
    let response = await ost.createTicket(ticket, options)
    expect(response.ticketId).to.be.a('number')
  })

  it('Try to create a ticket - should be ignored due to lock - return 423', async() => {
    const ticket = {
      name: 'Jane Doe',
      email: 'jane.doe@admiralcloud.com',
      subject: 'I need help',
      message: 'This is my bug report'
    }
    const options = {
      key: 'ticketLock',
      expires: 5
    }
    let response = await ost.createTicket(ticket, options)
    expect(response.status).to.eql(423)
  })

  it('Wait until lock is released', (done) => {
    setTimeout(done, 5000)
  })

  it('Should create a ticket after lock is released', async() => {
    const ticket = {
      name: 'Jane Doe',
      email: 'jane.doe@admiralcloud.com',
      subject: 'I need help',
      message: 'This is my bug report'
    }
    let response = await ost.createTicket(ticket)
    expect(response.ticketId).to.be.a('number')
  })

  it('Should create a ticket with custom fields - jobNumber should be in originalPayload', async() => {
    const ticket = {
      name: 'Jane Doe',
      email: 'jane.doe@admiralcloud.com',
      subject: 'I need help',
      message: 'This is my bug report',
      jobNumber: 'ABC-JobId-123',
      payloadCheck: true
    }
    const options = {
      fieldsToCheck: [
        { field: 'jobNumber', type: 'string', required: true },
      ]
    }
    let response = await ost.createTicket(ticket, options)
    expect(response.ticketId).to.be.a('number')
    const originalPayload = JSON.parse(Buffer.from(_.get(response, 'hash'), 'base64').toString('utf-8'))
    expect(originalPayload).to.have.property('jobNumber', ticket.jobNumber)

  })

  it('Should create a ticket with custom fields - jobNumber should be filtered out', async() => {
    const ticket = {
      name: 'Jane Doe',
      email: 'jane.doe@admiralcloud.com',
      subject: 'I need help',
      message: 'This is my bug report',
      jobNumber: 'ABC-JobId-123',
      payloadCheck: true
    }
    let response = await ost.createTicket(ticket)
    expect(response.ticketId).to.be.a('number')
    const originalPayload = JSON.parse(Buffer.from(_.get(response, 'hash'), 'base64').toString('utf-8'))
    expect(originalPayload).not.to.have.property('jobNumber')
  })


  it('Should create a ticket with DEBUG', async() => {
    const ticket = {
      name: 'Jane Doe',
      email: 'jane.doe@admiralcloud.com',
      subject: 'I need help',
      message: 'This is my bug report'
    }
    let response = await ost.createTicket(ticket, { debug: true })
    expect(response.ticketId).to.be.a('number')
    expect(response.debugMode).to.be.true
  })

  it('Should create a ticket with DEBUGMODE as init parameter', async() => {
    ost.init({
      debugMode: true
    })

    const ticket = {
      name: 'Jane Doe',
      email: 'jane.doe@admiralcloud.com',
      subject: 'I need help',
      message: 'This is my bug report'
    }
    let response = await ost.createTicket(ticket)
    expect(response.ticketId).to.be.a('number')
    expect(response.debugMode).to.be.true
  })

  it('Try to create a ticket at a non-existing server - should fail', async() => {
    ost.init({
      baseUrl: 'http://localhost:8071/',
      apiKey: 'abc-123',
      debugMode: false
    })

    const ticket = {
      name: 'Jane Doe',
      email: 'jane.doe@admiralcloud.com',
      subject: 'I need help',
      message: 'This is my bug report'
    }
    let response = await ost.createTicket(ticket)
    expect(response).to.be.have.property('code', 'ECONNREFUSED')
  })



 

})

.czConfig.js
module.exports = {
  repository: {
    baseUrl: 'https://github.com/',
    repoUrl: 'mmpro/ac-osticket/'
  },
  types: [
    {value: 'feat',     name: 'feat:     A new feature'},
    {value: 'fix',      name: 'fix:      A bug fix'},
    {value: 'docs',     name: 'docs:     Documentation only changes'},
    {value: 'style',    name: 'style:    Changes that do not affect the meaning of the code\n            (white-space, formatting, missing semi-colons, etc)'},
    {value: 'refactor', name: 'refactor: A code change that neither fixes a bug nor adds a feature'},
    {value: 'perf',     name: 'perf:     A code change that improves performance'},
    {value: 'test',     name: 'test:     Adding missing tests'},
    {value: 'chore',    name: 'chore:    Changes to the build process or auxiliary tools\n            and libraries such as documentation generation'},
    {value: 'revert',   name: 'revert:   Revert to a commit'},
    {value: 'WIP',      name: 'WIP:      Work in progress'}
  ],
  scopes: [
    {name: 'App' }
  ],
  allowCustomScopes: true,
  allowBreakingChanges: ['feat', 'fix']
};


index.js
const _ = require('lodash')
const axios = require('axios')

const keylock = require('ac-keylock')
const sanitizer = require('ac-sanitizer')

const osTicket = () => {

  // PRIVATE
  const apiBaseParams = {
    baseUrl: 'https://osticket.com/',
    headers: {
    //  'Content-Type': 'application/x-www-form-urlencoded',
      'x-api-key': 'abc-123',
    }
  }

  const apiCall = async function(params) {
    let apiParams = {
      method: _.get(params, 'method', 'post'),
      baseURL: _.get(apiBaseParams, 'baseUrl'),
      url: _.get(params, 'url', '/api/tickets.json'),
      headers: _.get(params, 'headers', _.get(apiBaseParams, 'headers')),
      data: _.get(params, 'data')
    }

    if (_.get(apiBaseParams, 'debugMode') || _.get(params, 'debug')) {
      const ticketId = Math.floor(Math.random() * 100000)
      return { ticketId, debugMode: true }
    }

    try {
      const apiResponse = await axios(apiParams)
      const response = {
        ticketId: _.get(apiResponse, 'data')
      }
      // for test mode
      if (_.get(apiResponse, 'headers.x-ac-payloadhash')) {
        _.set(response, 'hash', _.get(apiResponse, 'headers.x-ac-payloadhash'))
      }
      return response
    }
    catch(e) {
      console.log('OSTICKET | Payload | %j', _.pick(apiParams, ['baseURL', 'url', 'data']))
      console.log('OSTICKET | Error | %s | %j', _.get(e, 'code'), _.get(e, 'message'))
      return { code: _.get(e, 'code'), message: _.get(e, 'message') }
    }
  }


  // PUBLIC
  const init = async (params) => {
    _.set(apiBaseParams, 'baseUrl', _.get(params, 'baseUrl'))
    _.set(apiBaseParams, 'headers.x-api-key', _.get(params, 'apiKey'))
    if (_.get(params, 'apiSecret')) {
      _.set(apiBaseParams, 'headers.x-api-auth', _.get(params, 'apiSecret'))
    }
    if (_.has(params, 'debugMode')) {
      _.set(apiBaseParams, 'debugMode', _.get(params, 'debugMode'))
    }
    await keylock.init(_.get(params, 'keylock'))
  }
  
  const createTicket = async(data, options) => {
    let fields = [
      { field: 'email', type: 'email', required: true },
      { field: 'name', type: 'string', required: true },
      { field: 'subject', type: 'string', required: true },
      { field: 'message', type: 'string', required: true },
      { field: 'payloadCheck', type: 'boolean' },
    ]
    if (_.size(_.get(options, 'fieldsToCheck'))) {
      fields = _.concat(fields, _.get(options, 'fieldsToCheck'))
    }

    let fieldsToCheck = {
      params: data,
      fields
    }
    let check = sanitizer.checkAndSanitizeValues(fieldsToCheck)
    if (_.get(check, 'error')) return _.get(check, 'error')
    else data = _.get(check, 'params')

    let apiParams = {
      data,
      debug: _.get(options, 'debug')
    }
    
    if (!_.get(options, 'key')) {
      return await apiCall(apiParams)
    }
    else {
      const lockParams = {
        key: _.get(options, 'key'),
        value: _.get(options, 'value'),
        expires: _.get(options, 'expires')
      }
      let lock = await keylock.lockKey(lockParams)
      if (_.has(lock, 'status')) return { status: _.get(lock, 'status') }
      return await apiCall(apiParams)
    }
  }

  return {
    init,
    createTicket
  }
}

module.exports = osTicket()



